### PR TITLE
[Local Testing] SNOW-848935 Support VariantType, MapType and ArrayType

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
### **Please describe how your code solves the related issue.**

   Support creating dataframes with columns of VariantType, MapType and ArrayType by implementing conversion functions `to_variant`/`to_array`/`to_object`.

**Todo/Questions**

- Should we keep variant type as strings to align with the existing behavior?
- [ ] Support SubfieldString Expression

### CODE REVIEW

File: src/snowflake/snowpark/mock/functions.py [hunk 1]
```diff
@@ -633,9 +636,60 @@ def mock_endswith(expr1: ColumnEmulator, expr2: ColumnEmulator):
...
+    return res
+
+
+@patch("to_object")
+def mock_to_object(expr: ColumnEmulator):
+    if isinstance(expr.sf_type.datatype, (MapType,)):
+        res = res = copy(expr)
+    elif isinstance(expr.sf_type.datatype, VariantType):
+        res = expr.apply(
+            lambda x: try_convert(lambda y: y if isinstance(y, dict) else {y}, False, x)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/functions.py:677]: is `{y}` a set instead of a dict?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:677]: Thanks for catching this, the else case here should raise an exception.

File: src/snowflake/snowpark/mock/functions.py [hunk 2]
```diff
@@ -633,9 +636,60 @@ def mock_endswith(expr1: ColumnEmulator, expr2: ColumnEmulator):
...
     res.sf_type = ColumnType(StringType(), expr.sf_type.nullable)
     return res
 
+
+@patch("parse_json")
+def mock_parse_json(expr: ColumnEmulator):
+    if isinstance(expr.sf_type.datatype, StringType):
+        res = expr.apply(lambda x: try_convert(json.loads, False, x))
+    else:
+        res = copy(expr)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/functions.py:652]: we want a deep copy or shallow copy?
my assumption is that we want a deep copy.

also pandas provides a `copy` on df/series which does a deep copy: https://pandas.pydata.org/docs/reference/api/pandas.Series.copy.html#pandas-series-copy

File: src/snowflake/snowpark/mock/functions.py [hunk 3]
```diff
@@ -633,9 +636,60 @@ def mock_endswith(expr1: ColumnEmulator, expr2: ColumnEmulator):
...
+            lambda x: try_convert(lambda y: y if isinstance(y, dict) else {y}, False, x)
+        )
+    else:
+
+        def raise_exc():
+            raise SnowparkSQLException(
+                f"Invalid type {type(expr.sf_type.datatype)} parameter 'TO_OBJECT'"
+            )
+
+        res = expr.apply(lambda x: try_convert(raise_exc, False, x))
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/functions.py:686]: why we are applying a raise error function when trying to convert?
should we just raise error here?

File: tests/integ/scala/test_dataframe_suite.py [hunk 4]
```diff
@@ -1672,13 +1672,17 @@ def test_createDataFrame_with_given_schema_array_map_variant(session):
...
     schema = StructType(
         [StructField("array", ArrayType(None)), StructField("map", MapType(None, None))]
     )
     data = [Row([1, "\"'"], {"a": "\"'"})]
     df = session.create_dataframe(data, schema)
-    Utils.check_answer(df, [Row('[\n  1,\n  "\\"\'"\n]', '{\n  "a": "\\"\'"\n}')])
+    if not local_testing_mode:
+        Utils.check_answer(df, [Row('[\n  1,\n  "\\"\'"\n]', '{\n  "a": "\\"\'"\n}')])
+    else:
+        Utils.check_answer(df, data)
```
Comment (sfc-gh-aling) [inline tests/integ/scala/test_dataframe_suite.py:1685]: as we discussed, we shall update the output to json stringfy the object to align with the snowflake behavior

File: src/snowflake/snowpark/mock/functions.py [hunk 5]
```diff
@@ -633,9 +636,60 @@ def mock_endswith(expr1: ColumnEmulator, expr2: ColumnEmulator):
...
 def mock_upper(expr: ColumnEmulator):
     res = expr.apply(lambda x: x.upper())
     res.sf_type = ColumnType(StringType(), expr.sf_type.nullable)
     return res
 
+
+@patch("parse_json")
+def mock_parse_json(expr: ColumnEmulator):
+    if isinstance(expr.sf_type.datatype, StringType):
+        res = expr.apply(lambda x: try_convert(json.loads, False, x))
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py]: updated to add two module variable in function.py and connection.py to allow customizing encoder/decoder

File: src/snowflake/snowpark/mock/functions.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py]: using pandas.dataframe.copy

File: src/snowflake/snowpark/mock/functions.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py]: updated to raise error if not instance of dict

File: src/snowflake/snowpark/mock/functions.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py:700]: updated to raise error

File: tests/integ/scala/test_dataframe_suite.py (hunk 4 above)
Comment (sfc-gh-aling) [reply tests/integ/scala/test_dataframe_suite.py]: updated in connection.py to stringfy the json obj

File: local-testing-docs/README.md [hunk 6]
```diff
@@ -857,5 +857,10 @@ Please note that there will be unaware limitations, in this case please feel fre
   - Results of calculation regarding the values and data types
   - Columns names
   - Error experience
+- For data types Variant, Array and Object, basic capabilities of `parse_json`,
+  `to_array`, `to_object` and `to_variant` are provided via Python built-in JSONEncoder and
+  JSONDecoder which only support conversion of primitive Python data types. Advanced data conversion can be achieved by customizing
```
Comment (sfc-gh-stan) [inline local-testing-docs/README.md:862]: You could supply custom implementation of JSONEncoder and JSONDecoder by setting module level variables `snowflake.snowpark.mock.connection._CUSTOM_JSON_ENCODER` and `snowflake.snowpark.mock.functions._CUSTOM_JSON_DECODER`.

File: local-testing-docs/README.md [hunk 7]
```diff
@@ -857,5 +857,10 @@ Please note that there will be unaware limitations, in this case please feel fre
...
+  the `snowflake.snowpark.mock.connection._CUSTOM_JSON_ENCODER` and `snowflake.snowpark.mock.functions._CUSTOM_JSON_DECODER`.
+- Coercion is not supported
```
Comment (sfc-gh-stan) [inline local-testing-docs/README.md:864]: Implicit type casting (Coercion is not supported). In general, please avoid using passing in arguments of different types to functions.

File: local-testing-docs/README.md (hunk 6 above)
Comment (sfc-gh-aling) [reply local-testing-docs/README.md]: updated

File: local-testing-docs/README.md (hunk 7 above)
Comment (sfc-gh-aling) [reply local-testing-docs/README.md]: updated

File: src/snowflake/snowpark/mock/connection.py [hunk 8]
```diff
@@ -370,6 +377,24 @@ def execute(
 
         res = execute_mock_plan(plan)
         if isinstance(res, TableEmulator):
+            # stringfy the variant type in the result df
+            for col in res.columns:
+                # TODO: the first check is due to window function doesn't set res.sf_types[col]
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:382]: What's the first check? Is this outdated?

File: src/snowflake/snowpark/mock/functions.py [hunk 9]
```diff
@@ -231,7 +236,7 @@ def mock_listagg(column: ColumnEmulator, delimiter, is_distinct):
     columns_data = ColumnEmulator(column.unique()) if is_distinct else column
     return ColumnEmulator(
         data=delimiter.join([str(v) for v in columns_data.dropna()]),
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/functions.py:238]: Returns a string that includes all of the non-NULL input values, separated by the delimiter.

File: src/snowflake/snowpark/mock/connection.py (hunk 8 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:382]: oh yes this commenet it out of date, let me remove it

File: src/snowflake/snowpark/mock/functions.py (hunk 9 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py:238]: this is for commenting purpose?

File: src/snowflake/snowpark/mock/functions.py [hunk 10]
```diff
@@ -782,3 +787,66 @@ def mock_upper(expr: ColumnEmulator):
...
+            )
+        )
+    else:
+        res = expr.copy()
+    res.sf_type = ColumnType(VariantType(), expr.sf_type.nullable)
+    return res
+
+
+@patch("to_array")
+def mock_to_array(expr: ColumnEmulator):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/functions.py:808]:
```suggestion
def mock_to_array(expr: ColumnEmulator):
"""
[x] If the input is an ARRAY, or VARIANT containing an array value, the result is unchanged.

[ ] For NULL or (TODO:) a JSON null input, returns NULL.

[x] For any other value, the result is a single-element array containing this value.
"""
```

File: src/snowflake/snowpark/mock/functions.py (hunk 9 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:239]: Yes and I also think the implementation is not filtering the NULL values out 🤔

File: src/snowflake/snowpark/mock/functions.py [hunk 11]
```diff
@@ -782,3 +788,66 @@ def mock_upper(expr: ColumnEmulator):
...
+            lambda x: try_convert(lambda y: y if isinstance(y, list) else [y], False, x)
+        )
+    else:
+        res = expr.apply(lambda x: try_convert(lambda y: [y], False, x))
+    res.sf_type = ColumnType(ArrayType(), expr.sf_type.nullable)
+    return res
+
+
+@patch("to_object")
+def mock_to_object(expr: ColumnEmulator):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/functions.py:822]:
```suggestion
def mock_to_object(expr: ColumnEmulator):
"""
[x] For a VARIANT value containing an OBJECT, returns the OBJECT.

[ ] For NULL input, or for (TODO:) a VARIANT value containing only JSON null, returns NULL.

[x] For an OBJECT, returns the OBJECT itself.

[x] For all other input values, reports an error.
"""
```

File: src/snowflake/snowpark/mock/plan.py [hunk 12]
```diff
@@ -1112,6 +1122,10 @@ def calculate_expression(
             return _MOCK_FUNCTION_IMPLEMENTATION_MAP["to_double"](
                 column, try_cast=exp.try_
             )
+        elif isinstance(exp.to, MapType):
+            return _MOCK_FUNCTION_IMPLEMENTATION_MAP["to_object"](column)
+        elif isinstance(exp.to, ArrayType):
+            return _MOCK_FUNCTION_IMPLEMENTATION_MAP["to_array"](column)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1128]: Let's also add to `to_variant` too.

File: src/snowflake/snowpark/mock/plan.py [hunk 13]
```diff
@@ -1299,15 +1314,22 @@ def calculate_expression(
...
-                        )
+                    sub_window_res = calculate_expression(
+                        window_function.expr,
+                        w.iloc[offset_idx],
+                        analyzer,
+                        expr_to_alias,
+                        keep_literal=True,
                     )
+                    # we use the whole frame to calculate the type
+                    calculated_sf_type = calculate_expression(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1325]: This could be moved outside the loop.

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/plan.py:1325]: We should raise an exception for coercion if the two sf_types are conflicting.

File: src/snowflake/snowpark/mock/plan.py [hunk 14]
```diff
@@ -1338,11 +1360,15 @@ def calculate_expression(
...
                         res_cols.append(target_expr)
-
             res_col = ColumnEmulator(
                 data=res_cols, dtype=object
             )  # dtype=object prevents implicit converting None to Nan
             res_col.index = res_index
+            res_col.sf_type = (
+                calculated_sf_type
+                if calculated_sf_type
+                else ColumnType(NullType(), True)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1370]: We should not use `NullType` here.

File: src/snowflake/snowpark/mock/plan.py [hunk 15]
```diff
@@ -1373,7 +1399,15 @@ def calculate_expression(
                     else:
                         res_cols.append(None)
             res_col = ColumnEmulator(
-                data=res_cols, dtype=object
+                data=res_cols,
+                dtype=object,
+                sf_type=calculate_expression(
+                    window_function.expr,
+                    windows[0],
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1408]: We could use `windows[0]` but I also think we could use input_data directly. All the required columns should be present.

File: src/snowflake/snowpark/mock/plan.py [hunk 16]
```diff
@@ -1406,14 +1442,42 @@ def calculate_expression(
...
             return res_col.sort_index()
         else:
             raise NotImplementedError(
                 f"[Local Testing] Window Function {window_function} is not implemented."
             )
+    elif isinstance(exp, SubfieldString):
+        col = calculate_expression(exp.child, input_data, analyzer, expr_to_alias)
+        field = str(exp.field)
+        # in snowflake, two consecutive single quotes means escaping single quote
+        field = field.replace("''", "'")
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1465]: Question: Where does the quote get escaped?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py:1465]: singe quote is escaped if it's a string constant: https://docs.snowflake.com/en/sql-reference/data-types-text#single-quoted-string-constants

in this PR, it's the test
```python
    def variant2(cls, session: "Session") -> DataFrame:
        return session.sql(
            """
            select parse_json(column1) as src
            from values
            ('{
                "date with '' and ." : "2017-04-28",
                "salesperson" : {
                  "id": "55",
                  "name": "Frank Beasley"
                },
                "customer" : [
                  {"name": "Joyce Ridgely", "phone": "16504378889", "address": "San Francisco, CA"}
                ],
                "vehicle" : [
                  {"make": "Honda", "extras":["ext warranty", "paint protection"]}
                ]
            }')
            """
```

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py:1465]: probably we need to apply this escape logic when parsing the string literal

File: src/snowflake/snowpark/mock/plan.py [hunk 17]
```diff
@@ -1373,7 +1401,15 @@ def calculate_expression(
...
             res_col = ColumnEmulator(
-                data=res_cols, dtype=object
+                data=res_cols,
+                dtype=object,
+                sf_type=calculate_expression(
+                    window_function.expr,
+                    windows[0],
+                    analyzer,
+                    expr_to_alias,
+                    keep_literal=True,
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1411]: Remove

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 18]
```diff
@@ -285,6 +286,13 @@ def __init__(self, *args, **kwargs) -> NoReturn:
         sf_type = kwargs.pop("sf_type", None)
         super().__init__(*args, **kwargs)
         self.sf_type: ColumnType = sf_type
+        # record which rows should be marked as null instead of None
+        # snowflake SubfieldString has this behavior
+        # suppose there are two Variant objects in table "v": 1. { "a": None } 2. None
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:291]: I understand this is to align with live connection behavior. But I think this is a buggy behavior, let's try to fix it main and revert this in local testing.

File: tests/mock/test_agg.py [hunk 19]
```diff
@@ -110,7 +111,7 @@ def mock_approx_percentile(
     ) -> ColumnEmulator:
         assert column.tolist() == [10.0, 20.0, 25.0, 30.0]
         assert percentile == 0.5
-        return ColumnEmulator(data=123)
+        return ColumnEmulator(data=123, sf_type=ColumnType(DoubleType(), False))
```
Comment (sfc-gh-stan) [inline tests/mock/test_agg.py:114]: Let's not enforce non-nullable unless we are asserting it.

File: src/snowflake/snowpark/mock/plan.py (hunk 14 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py:1410]: I keep the NullType here just for defensive coding for the case if all the windows are return None

File: src/snowflake/snowpark/mock/snowflake_data_type.py (hunk 18 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/snowflake_data_type.py:291]: JIRA created: SNOW-960190

File: src/snowflake/snowpark/mock/plan.py (hunk 15 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: updated

File: tests/mock/test_agg.py (hunk 19 above)
Comment (sfc-gh-aling) [reply tests/mock/test_agg.py:114]: ColumnType is a namedtuple requiring the positional argument, I can flip it to true but I don't think it makes a different here..

File: src/snowflake/snowpark/mock/plan.py (hunk 17 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: removed

File: src/snowflake/snowpark/mock/plan.py (hunk 13 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: updated to raise error if cur type != pre type

File: src/snowflake/snowpark/mock/plan.py [hunk 20]
```diff
@@ -1289,25 +1306,72 @@ def calculate_expression(
...
+                    # we use the whole frame to calculate the type
+                    cur_windows_sf_type = calculate_expression(
+                        window_function.expr,
+                        w,
+                        analyzer,
+                        expr_to_alias,
+                    ).sf_type
+                    if not calculated_sf_type:
+                        calculated_sf_type = cur_windows_sf_type
+                        if not calculated_sf_type:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1352]: not `NullType()`  evaluates to True?

File: src/snowflake/snowpark/mock/plan.py [hunk 21]
```diff
@@ -1289,25 +1306,72 @@ def calculate_expression(
...
+                            # this is defensive code, if calculated_sf_type is not calculated yet or calculated
+                            # being None before, we don't set calculated_sf_type to NullType
+                            # because there could be chances that the first couple windows are indeed
+                            # returning None, but the windows afterwards will generate result of non-None type
+                            # and after we calculate all the windows, if still calculated_sf_type is not set
+                            # then we should be safely set the calculated_sf_type to None
+                            if not isinstance(cur_windows_sf_type.datatype, NullType):
+                                calculated_sf_type = cur_windows_sf_type
+                        elif not isinstance(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1361]: We can just use == here.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**